--- a/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c
+++ b/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c
@@ -173,63 +173,6 @@ int ext_if_add(struct extdev_entry *ext_
 	return len;
 }
 
-static void foe_clear_ethdev_bind_entries(struct net_device *dev)
-{
-	struct net_device *master_dev = dev;
-	const struct dsa_port *dp;
-	struct foe_entry *entry;
-	struct mtk_mac *mac;
-	bool match_dev = false;
-	int port_id, gmac;
-	u32 i, hash_index;
-	u32 dsa_tag;
-	u32 total = 0;
-	/* Get the master device if the device is slave device */
-	port_id = hnat_dsa_get_port(&master_dev);
-	mac = netdev_priv(master_dev);
-	gmac = HNAT_GMAC_FP(mac->id);
-	if (gmac < 0)
-		return;
-	if (port_id >= 0) {
-		dp = dsa_port_from_netdev(dev);
-		if (IS_ERR(dp))
-			return;
-		if (0)
-			dsa_tag = port_id + BIT(11);
-		else
-			dsa_tag = BIT(port_id);
-	}
-	for (i = 0; i < CFG_PPE_NUM; i++) {
-		for (hash_index = 0; hash_index < hnat_priv->foe_etry_num; hash_index++) {
-			entry = hnat_priv->foe_table_cpu[i] + hash_index;
-			if (!entry_hnat_is_bound(entry))
-				continue;
-			match_dev = (IS_IPV4_GRP(entry)) ? entry->ipv4_hnapt.iblk2.dp == gmac :
-							   entry->ipv6_5t_route.iblk2.dp == gmac;
-			if (match_dev && port_id >= 0) {
-				if (0) {
-					match_dev = (IS_IPV4_GRP(entry)) ?
-						entry->ipv4_hnapt.vlan1 == dsa_tag :
-						entry->ipv6_5t_route.vlan1 == dsa_tag;
-				} else {
-					match_dev = (IS_IPV4_GRP(entry)) ?
-						!!(entry->ipv4_hnapt.etype & dsa_tag) :
-						!!(entry->ipv6_5t_route.etype & dsa_tag);
-				}
-			}
-			if (match_dev) {
-				entry->bfib1.state = INVALID;
-				entry->bfib1.time_stamp =
-					readl((hnat_priv->fe_base + 0x0010)) & 0xFF;
-				total++;
-			}
-		}
-	}
-	/* clear HWNAT cache */
-	if (total > 0)
-		hnat_cache_ebl(1);
-}
-
 int ext_if_del(struct extdev_entry *ext_entry)
 {
 	int i, j;
@@ -426,15 +369,15 @@ int nf_hnat_netdevice_event(struct notif
 
 	switch (event) {
 	case NETDEV_UP:
-
-		if (!hnat_priv->guest_en && dev->name) {
+		if (!hnat_priv->guest_en) {
 			if (!strcmp(dev->name, "ra1") || !strcmp(dev->name, "rax1") || !strcmp(dev->name, "rai1")) {
 				break;
+			}
 		}
 
 		gmac_ppe_fwd_enable(dev);
 
-		extif_set_dev(dev,1);
+		extif_set_dev(dev);
 
 		break;
 	case NETDEV_CHANGE:
@@ -3330,7 +3273,7 @@ void mtk_ppe_dev_register_hook(struct ne
 	for (i = 1; i < MAX_IF_NUM; i++) {
 		if (!hnat_priv->wifi_hook_if[i]) {
 			if (find_extif_from_devname(dev->name)) {
-				extif_set_dev(dev,0);
+				extif_set_dev(dev);
 				goto add_wifi_hook_if;
 			}
 
