diff -Nuar linux-6.12.94-old/drivers/net/ethernet/mediatek/mtk_hnat/hnat_debugfs.c linux-6.12.94-new/drivers/net/ethernet/mediatek/mtk_hnat/hnat_debugfs.c
--- linux-6.12.94-old/drivers/net/ethernet/mediatek/mtk_hnat/hnat_debugfs.c	2026-07-06 16:20:03.214373453 +0800
+++ linux-6.12.94-new/drivers/net/ethernet/mediatek/mtk_hnat/hnat_debugfs.c	2026-07-06 16:20:33.827517347 +0800
@@ -1354,6 +1354,8 @@
 				counter = acct->counter;
 				atomic64_add(bytes, &counter[dir].bytes);
 				atomic64_add(packets, &counter[dir].packets);
+				atomic64_set(&counter[dir].diff_bytes, bytes);
+				atomic64_set(&counter[dir].diff_packets, packets);
 			}
 
 			/* avoid compiler warnings */
diff -Nuar linux-6.12.94-old/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c linux-6.12.94-new/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c
--- linux-6.12.94-old/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c	2026-07-06 16:20:03.214373453 +0800
+++ linux-6.12.94-new/drivers/net/ethernet/mediatek/mtk_hnat/hnat_nf_hook.c	2026-07-06 16:20:33.827517347 +0800
@@ -1256,6 +1256,67 @@
 		cmpxchg(is_ingress ? &acct->iif : &acct->oif, 0, skb->dev->ifindex);
 }
 
+/* update hnat count to nf_conntrack and iptables by keepalive */
+static void mtk_hnat_nf_update(struct sk_buff *skb)
+{
+	struct nf_conn *ct;
+	struct nf_conn_acct *acct;
+	struct nf_conn_counter *counter;
+	enum ip_conntrack_info ctinfo;
+	struct hnat_accounting diff;
+
+	if (skb->protocol == htons(ETH_P_IPV6) && !hnat_priv->ipv6_en)
+		return;
+
+	if (skb_hnat_alg(skb) || unlikely(!is_magic_tag_valid(skb) ||
+					  !IS_SPACE_AVAILABLE_HEAD(skb)))
+		return;
+
+	if (unlikely(!skb_mac_header_was_set(skb)))
+		return;
+
+	if (unlikely(!skb_hnat_is_hashed(skb)))
+		return;
+
+	if (unlikely(skb->mark == HNAT_EXCEPTION_TAG))
+		return;
+
+	ct = nf_ct_get(skb, &ctinfo);
+	if (ct) {
+		if (!hnat_get_count(hnat_priv, skb_hnat_ppe(skb),
+				    skb_hnat_entry(skb), &diff))
+			return;
+
+		acct = nf_conn_acct_find(ct);
+		if (acct) {
+			counter = acct->counter;
+			atomic64_add(diff.packets,
+				     &counter[CTINFO2DIR(ctinfo)].packets);
+			atomic64_add(diff.bytes,
+				     &counter[CTINFO2DIR(ctinfo)].bytes);
+			atomic64_set(&counter[CTINFO2DIR(ctinfo)].diff_packets,
+				     diff.packets);
+			atomic64_set(&counter[CTINFO2DIR(ctinfo)].diff_bytes,
+				     diff.bytes);
+		}
+	}
+}
+
+static unsigned int
+mtk_hnat_nf_conntrack(void *priv, struct sk_buff *skb,
+		      const struct nf_hook_state *state)
+{
+	if (!skb)
+		return NF_DROP;
+
+	if (unlikely(skb_hnat_reason(skb) == HIT_BIND_KEEPALIVE_DUP_OLD_HDR)) {
+		if (hnat_priv->data->per_flow_accounting && hnat_priv->nf_stat_en)
+			mtk_hnat_nf_update(skb);
+	}
+
+	return NF_ACCEPT;
+}
+
 static unsigned int
 mtk_hnat_ipv6_nf_pre_routing(void *priv, struct sk_buff *skb,
 			     const struct nf_hook_state *state)
@@ -4366,6 +4427,18 @@
 
 static struct nf_hook_ops mtk_hnat_nf_ops[] __read_mostly = {
 	{
+		.hook = mtk_hnat_nf_conntrack,
+		.pf = NFPROTO_IPV4,
+		.hooknum = NF_INET_PRE_ROUTING,
+		.priority = NF_IP_PRI_MANGLE - 1,
+	},
+	{
+		.hook = mtk_hnat_nf_conntrack,
+		.pf = NFPROTO_IPV6,
+		.hooknum = NF_INET_PRE_ROUTING,
+		.priority = NF_IP_PRI_MANGLE - 1,
+	},
+	{
 		.hook = mtk_hnat_ipv4_nf_pre_routing,
 		.pf = NFPROTO_IPV4,
 		.hooknum = NF_INET_PRE_ROUTING,
diff -Nuar linux-6.12.94-old/net/ipv4/netfilter/ip_tables.c linux-6.12.94-new/net/ipv4/netfilter/ip_tables.c
--- linux-6.12.94-old/net/ipv4/netfilter/ip_tables.c	2026-07-06 16:20:01.750967793 +0800
+++ linux-6.12.94-new/net/ipv4/netfilter/ip_tables.c	2026-07-06 16:20:33.827517347 +0800
@@ -25,6 +25,8 @@
 #include <linux/netfilter/x_tables.h>
 #include <linux/netfilter_ipv4/ip_tables.h>
 #include <net/netfilter/nf_log.h>
+#include <net/netfilter/nf_conntrack.h>
+#include <net/netfilter/nf_conntrack_acct.h>
 #include "../../netfilter/xt_repldata.h"
 
 MODULE_LICENSE("GPL");
@@ -271,6 +273,26 @@
 	return true;
 }
 
+static struct __attribute__((__packed__)) hnat_desc {
+	u32 entry : 18;
+#if defined(CONFIG_MEDIATEK_NETSYS_RX_V2)
+	u32 fill : 4;
+#endif
+	u32 crsn : 5;
+	u32 sport : 4;
+	u32 alg : 1;
+	u32 iface : 8;
+	u32 ppe : 1;
+	u32 filled : 3;
+	u32 magic_tag_protect : 16;
+	u32 wdmaid : 8;
+	u32 rxid : 2;
+	u32 wcid : 30;
+	u32 bssid : 32;
+} __maybe_unused __hnat_desc;
+#define HIT_BIND_KEEPALIVE_DUP_OLD_HDR 0x15
+#define skb_hnat_reason(skb) (((struct hnat_desc *)(skb->head))->crsn)
+
 /* Returns one of the generic firewall policies, like NF_ACCEPT. */
 unsigned int
 ipt_do_table(void *priv,
@@ -291,6 +313,11 @@
 	struct xt_action_param acpar;
 	unsigned int addend;
 
+	struct nf_conn *ct;
+	struct nf_conn_acct *acct;
+	struct nf_conn_counter *conn_counter;
+	enum ip_conntrack_info ctinfo;
+
 	/* Initialization */
 	WARN_ON(!(table->valid_hooks & (1 << hook)));
 	local_bh_disable();
@@ -304,6 +331,20 @@
 
 		counter = xt_get_this_cpu_counter(&e->counters);
 		ADD_COUNTER(*counter, skb->len, 1);
+
+		if (unlikely(skb_hnat_reason(skb) == HIT_BIND_KEEPALIVE_DUP_OLD_HDR)) {
+			ct = nf_ct_get(skb, &ctinfo);
+			if (ct) {
+				acct = nf_conn_acct_find(ct);
+				if (acct) {
+					conn_counter = acct->counter;
+					ADD_COUNTER(*counter,
+						atomic64_read(&conn_counter[CTINFO2DIR(ctinfo)].diff_bytes),
+						atomic64_read(&conn_counter[CTINFO2DIR(ctinfo)].diff_packets));
+				}
+			}
+		}
+
 		local_bh_enable();
 		return verdict;
 	}
@@ -360,6 +401,19 @@
 		counter = xt_get_this_cpu_counter(&e->counters);
 		ADD_COUNTER(*counter, skb->len, 1);
 
+		if (unlikely(skb_hnat_reason(skb) == HIT_BIND_KEEPALIVE_DUP_OLD_HDR)) {
+			ct = nf_ct_get(skb, &ctinfo);
+			if (ct) {
+				acct = nf_conn_acct_find(ct);
+				if (acct) {
+					conn_counter = acct->counter;
+					ADD_COUNTER(*counter,
+						atomic64_read(&conn_counter[CTINFO2DIR(ctinfo)].diff_bytes),
+						atomic64_read(&conn_counter[CTINFO2DIR(ctinfo)].diff_packets));
+				}
+			}
+		}
+
 		t = ipt_get_target_c(e);
 		WARN_ON(!t->u.kernel.target);
 
diff -Nuar linux-6.12.94-old/net/ipv6/netfilter/ip6_tables.c linux-6.12.94-new/net/ipv6/netfilter/ip6_tables.c
--- linux-6.12.94-old/net/ipv6/netfilter/ip6_tables.c	2026-06-19 19:42:39.000000000 +0800
+++ linux-6.12.94-new/net/ipv6/netfilter/ip6_tables.c	2026-07-06 16:20:33.827517347 +0800
@@ -29,6 +29,8 @@
 #include <linux/netfilter_ipv6/ip6_tables.h>
 #include <linux/netfilter/x_tables.h>
 #include <net/netfilter/nf_log.h>
+#include <net/netfilter/nf_conntrack.h>
+#include <net/netfilter/nf_conntrack_acct.h>
 #include "../../netfilter/xt_repldata.h"
 
 MODULE_LICENSE("GPL");
@@ -243,6 +245,26 @@
 	return (void *)entry + entry->next_offset;
 }
 
+static struct __attribute__((__packed__)) hnat_desc {
+	u32 entry : 18;
+#if defined(CONFIG_MEDIATEK_NETSYS_RX_V2)
+	u32 fill : 4;
+#endif
+	u32 crsn : 5;
+	u32 sport : 4;
+	u32 alg : 1;
+	u32 iface : 8;
+	u32 ppe : 1;
+	u32 filled : 3;
+	u32 magic_tag_protect : 16;
+	u32 wdmaid : 8;
+	u32 rxid : 2;
+	u32 wcid : 30;
+	u32 bssid : 32;
+} __maybe_unused __hnat_desc;
+#define HIT_BIND_KEEPALIVE_DUP_OLD_HDR 0x15
+#define skb_hnat_reason(skb) (((struct hnat_desc *)(skb->head))->crsn)
+
 /* Returns one of the generic firewall policies, like NF_ACCEPT. */
 unsigned int
 ip6t_do_table(void *priv, struct sk_buff *skb,
@@ -261,6 +283,11 @@
 	struct xt_action_param acpar;
 	unsigned int addend;
 
+	struct nf_conn *ct;
+	struct nf_conn_acct *acct;
+	struct nf_conn_counter *conn_counter;
+	enum ip_conntrack_info ctinfo;
+
 	/* Initialization */
 	stackidx = 0;
 	indev = state->in ? state->in->name : nulldevname;
@@ -320,6 +347,19 @@
 		counter = xt_get_this_cpu_counter(&e->counters);
 		ADD_COUNTER(*counter, skb->len, 1);
 
+		if (unlikely(skb_hnat_reason(skb) == HIT_BIND_KEEPALIVE_DUP_OLD_HDR)) {
+			ct = nf_ct_get(skb, &ctinfo);
+			if (ct) {
+				acct = nf_conn_acct_find(ct);
+				if (acct) {
+					conn_counter = acct->counter;
+					ADD_COUNTER(*counter,
+						atomic64_read(&conn_counter[CTINFO2DIR(ctinfo)].diff_bytes),
+						atomic64_read(&conn_counter[CTINFO2DIR(ctinfo)].diff_packets));
+				}
+			}
+		}
+
 		t = ip6t_get_target_c(e);
 		WARN_ON(!t->u.kernel.target);
 
