--- a/regs/reg_v2/warp_hw_ccif.c
+++ b/regs/reg_v2/warp_hw_ccif.c
@@ -76,7 +76,7 @@ warp_ccif_isr_ctrl(void *hw, bool enable
 		warp_io_write32((struct ccif_entry *) hw, WOX_AP2WO_MCU_CCIF4_CPUA_PCCIF_IRQ0_MASK_ADDR, 0);
 }
 
-const static u32 ccif_dummy_table[] = {
+static const u32 ccif_dummy_table[] = {
 	WOX_AP2WO_MCU_CCIF4_CPUA_PCCIF_DUMMY1_ADDR,
 	WOX_AP2WO_MCU_CCIF4_CPUA_PCCIF_DUMMY2_ADDR,
 	WOX_AP2WO_MCU_CCIF4_CPUA_PCCIF_DUMMY3_ADDR,
@@ -87,7 +87,7 @@ const static u32 ccif_dummy_table[] = {
 	WOX_AP2WO_MCU_CCIF4_CPUA_PCCIF_DUMMY8_ADDR
 };
 
-const static u32 ccif_shadow_table[] = {
+static const u32 ccif_shadow_table[] = {
 	WOX_AP2WO_MCU_CCIF4_CPUB2CPUA_SHADOW1_ADDR,
 	WOX_AP2WO_MCU_CCIF4_CPUB2CPUA_SHADOW2_ADDR,
 	WOX_AP2WO_MCU_CCIF4_CPUB2CPUA_SHADOW3_ADDR,
--- a/regs/reg_v3/warp_hw_ccif.c
+++ b/regs/reg_v3/warp_hw_ccif.c
@@ -76,7 +76,7 @@ warp_ccif_isr_ctrl(void *hw, bool enable
 		warp_io_write32((struct ccif_entry *) hw, WOX_AP2WO_MCU_CCIF4_CPUA_PCCIF_IRQ0_MASK_OFFSET, 0);
 }
 
-const static u32 ccif_dummy_table[] = {
+static const u32 ccif_dummy_table[] = {
 	WOX_AP2WO_MCU_CCIF4_CPUA_PCCIF_DUMMY1_OFFSET,
 	WOX_AP2WO_MCU_CCIF4_CPUA_PCCIF_DUMMY2_OFFSET,
 	WOX_AP2WO_MCU_CCIF4_CPUA_PCCIF_DUMMY3_OFFSET,
@@ -87,7 +87,7 @@ const static u32 ccif_dummy_table[] = {
 	WOX_AP2WO_MCU_CCIF4_CPUA_PCCIF_DUMMY8_OFFSET
 };
 
-const static u32 ccif_shadow_table[] = {
+static const u32 ccif_shadow_table[] = {
 	WOX_AP2WO_MCU_CCIF4_CPUB2CPUA_SHADOW1_OFFSET,
 	WOX_AP2WO_MCU_CCIF4_CPUB2CPUA_SHADOW2_OFFSET,
 	WOX_AP2WO_MCU_CCIF4_CPUB2CPUA_SHADOW3_OFFSET,
